### PR TITLE
Segmentation training update

### PR DESCRIPTION
**Pull request by nilsleh**
This PR adds more functionality to Segmentation by doing the following:

- Implement a feature pyramid style decoder that is lightweight and can be trained fast, for transformer architectures, we also have Isaac's recommendation for DPT
- add encoder layer_names from which to extract the segmentation features to be used
- add optional optuna bayesian hparam search for learning rate and batch size: procedure is similar to logistic regression in the sense that best hparams are computed based on validation performance, and then best hparams are rerun on merged train+val dataset and then evaluated on final test score that is written to results output file
- now added caching mechanism which extracts the features across layers for input imagery. These features are cached once and then moved to GPU which gives a huge speed up for training the segmentation head

I have some experiments running to test performance.

### CODE REVIEW

**Comment by nilsleh** [comment]
These are the results for running hparam optimization for batch size and learning rate over a couple different models on two datasets


|dataset|method|metric_name|metric_value       |ci_lower|ci_upper|feature_dim|best_c|best_lr               |best_batch_size|n_train|n_val|n_test|seed|model                                    |name                 |normalization|image_size|interpolation|partition|c_range_start|c_range_stop|c_range_num|merge_val|bootstrap|
|-------|------|-----------|-------------------|--------|--------|-----------|------|----------------------|---------------|-------|-----|------|----|-----------------------------------------|---------------------|-------------|----------|-------------|---------|-------------|------------|-----------|---------|---------|
|flair2 |fpn   |mIoU       |0.4447839856147766 |0.0     |0.0     |960        |      |0.0005059803874660431 |16             |4049   |1022 |3022  |0   |torchgeo_bench.models.TimmPatchBenchModel|resnet18             |mean_stdev   |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|caffe  |fpn   |mIoU       |0.5451910495758057 |0.0     |0.0     |960        |      |0.0005059803874660431 |16             |4000   |1000 |2000  |0   |torchgeo_bench.models.TimmPatchBenchModel|resnet18             |mean_stdev   |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|flair2 |fpn   |mIoU       |0.46486544609069824|0.0     |0.0     |3840       |      |0.00044303752452182633|16             |4049   |1022 |3022  |0   |torchgeo_bench.models.TimmPatchBenchModel|resnet50             |mean_stdev   |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|caffe  |fpn   |mIoU       |0.5473743677139282 |0.0     |0.0     |3840       |      |0.0005059803874660431 |16             |4000   |1000 |2000  |0   |torchgeo_bench.models.TimmPatchBenchModel|resnet50             |mean_stdev   |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|flair2 |fpn   |mIoU       |0.46861737966537476|0.0     |0.0     |672        |      |0.008626905220714904  |16             |4049   |1022 |3022  |0   |torchgeo_bench.models.TimmPatchBenchModel|efficientnet_b2      |mean_stdev   |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|caffe  |fpn   |mIoU       |0.5349847078323364 |0.0     |0.0     |672        |      |0.0005059803874660431 |16             |4000   |1000 |2000  |0   |torchgeo_bench.models.TimmPatchBenchModel|efficientnet_b2      |mean_stdev   |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|flair2 |fpn   |mIoU       |0.4230736494064331 |0.0     |0.0     |736        |      |0.008626905220714904  |16             |4049   |1022 |3022  |0   |torchgeo_bench.models.TimmPatchBenchModel|mobilenetv3_small_100|mean_stdev   |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|caffe  |fpn   |mIoU       |0.5093039870262146 |0.0     |0.0     |736        |      |0.00044303752452182633|16             |4000   |1000 |2000  |0   |torchgeo_bench.models.TimmPatchBenchModel|mobilenetv3_small_100|mean_stdev   |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|flair2 |fpn   |mIoU       |0.45286816358566284|0.0     |0.0     |1440       |      |0.00044303752452182633|16             |4049   |1022 |3022  |0   |torchgeo_bench.models.TimmPatchBenchModel|convnext_small       |mean_stdev   |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|caffe  |fpn   |mIoU       |0.5314532518386841 |0.0     |0.0     |1440       |      |0.008626905220714904  |16             |4000   |1000 |2000  |0   |torchgeo_bench.models.TimmPatchBenchModel|convnext_small       |mean_stdev   |224       |bilinear     |default  |-7           |2           |20         |True     |200      |

**Comment by nilsleh** [comment]
Screenshot from 2026-04-13 15-39-57

FLAIR2: performance ranges from 36.7 to 49.8 across foundation models.


Screenshot from 2026-04-13 15-39-39

CaFfe: performance ranges from 53.7 to 67.3 across foundation models.

For reference, these are the values from GeoBenchV2 experiments, although the eval setup is not exactly the same, i.e. different segmentation head, different final training set of merged train + val.

**Comment by nilsleh** [comment]
Here are some ViT results:

|dataset|method|metric_name|metric_value       |ci_lower|ci_upper|feature_dim|best_c|best_lr               |best_batch_size|n_train|n_val|n_test|seed|model                                      |name                    |normalization|image_size|interpolation|partition|c_range_start|c_range_stop|c_range_num|merge_val|bootstrap|
|-------|------|-----------|-------------------|--------|--------|-----------|------|----------------------|---------------|-------|-----|------|----|-------------------------------------------|------------------------|-------------|----------|-------------|---------|-------------|------------|-----------|---------|---------|
|caffe  |dpt   |mIoU       |0.5914007425308228 |0.0     |0.0     |3072       |      |0.0005059803874660431 |16             |4000   |1000 |2000  |0   |torchgeo_bench.models.TorchGeoDOFABench    |tgeo_dofa_base          |none         |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|flair2 |dpt   |mIoU       |0.4700523018836975 |0.0     |0.0     |3072       |      |0.00044303752452182633|16             |4049   |1022 |3022  |0   |torchgeo_bench.models.TorchGeoDOFABench    |tgeo_dofa_base          |none         |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|caffe  |dpt   |mIoU       |0.5410823225975037 |0.0     |0.0     |1024       |      |0.0005059803874660431 |8              |4000   |1000 |2000  |0   |torchgeo_bench.models.SAM3EncoderBench     |sam3_encoder            |mean_stdev   |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|flair2 |dpt   |mIoU       |0.4535556733608246 |0.0     |0.0     |1024       |      |0.006823493012435794  |8              |4049   |1022 |3022  |0   |torchgeo_bench.models.SAM3EncoderBench     |sam3_encoder            |mean_stdev   |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|caffe  |dpt   |mIoU       |0.4179384112358093 |0.0     |0.0     |4096       |      |0.00020547625125911338|16             |4000   |1000 |2000  |0   |torchgeo_bench.models.TorchGeoScaleMAEBench|tgeo_scalemae_large_fmow|none         |224       |bilinear     |default  |-7           |2           |20         |True     |200      |
|flair2 |dpt   |mIoU       |0.25675690174102783|0.0     |0.0     |4096       |      |0.00014135935551752297|8              |4049   |1022 |3022  |0   |torchgeo_bench.models.TorchGeoScaleMAEBench|tgeo_scalemae_large_fmow|none         |224       |bilinear     |default  |-7           |2           |20         |True     |200      |

**Comment by calebrob6** [comment]
@nilsleh do these results make sense to you?

**Comment by nilsleh** [comment]
They seem a bit low to me, I was expecting a bit more. But I will open up the overfit PR logic, which I think will be a decent pipeline test. I think it would definitely be good to get more experiments and more eyes on this pipeline and check whether I have made some mistakes. But the rest of this PR is now broken into the other two features that I hope make the segmentation pipeline more investigative:
- #29 
- #31
